### PR TITLE
adds TOTP generation, validation, and backup codes

### PR DESCRIPTION
**Pull request by tristancalderbank**
:tickets: **Ticket(s)**:  regarding #417

---

## :construction_worker: Changes

- Adds TOTP secret generation/validation
- Adds backup code generation using same format as Github (e.g. "b2e03-ffbcf")
- TOTP secret generation parameters are compatible with Google Authenticator

## :flashlight: Testing Instructions

See: totp_test.go

### CODE REVIEW

**Comment by bobheadxi** [comment]
@tristancalderbank there's a fix in master that stops the builds from failing - if you update your branch travis should pass 👍

**Comment by tristancalderbank** [comment]
kk should be fixed

File: daemon/inertiad/crypto/totp_test.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+	// generate TOTP before current window (invalid)
+	badTime := currentTime.Add(time.Duration(-(TotpPeriod * 2) * time.Second))
+	code, err = totp.GenerateCode(key.Secret(), badTime)
+	assert.False(t, validatePasscode(code, key.Secret()))
+
+	// generate TOTP after current window (invalid)
+	badTime = currentTime.Add(time.Duration((TotpPeriod * 2) * time.Second))
+	code, err = totp.GenerateCode(key.Secret(), badTime)
+	assert.False(t, validatePasscode(code, key.Secret()))
```
Comment (bobheadxi) [inline daemon/inertiad/crypto/totp_test.go]: a table-driven test would be nice here (https://github.com/golang/go/wiki/TableDrivenTests), its not very well enforced throughout the codebase however so I guess its not super necessary

File: daemon/inertiad/crypto/totp.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+// Generates secret key object which can be turned into string or image
+func generateSecretKey(accountName string) (*otp.Key, error) {
+	key,err := totp.Generate(totp.GenerateOpts{
+		Issuer: TotpIssuerName,
+		AccountName: accountName,
+		Period: TotpPeriod,
+		SecretSize: TotpSecretSize,
+		Digits: TotpDigits,
```
Comment (bobheadxi) [inline daemon/inertiad/crypto/totp.go]: can you `go fmt`? it should complain about this formatting

**Comment by tristancalderbank** [comment]
sure, hows this

File: .gitignore [hunk 3]
```diff
@@ -40,3 +40,6 @@ public/
 
 # Exceptions
 !inertia.go
+
+# JetBrains
+.idea
```
Comment (bfbachmann) [inline .gitignore]: Oh my god dude. Stop using this malware.

File: daemon/inertiad/crypto/totp.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/pquerna/otp"
+	"github.com/pquerna/otp/totp"
+)
+
+const (
+	TotpIssuerName       = "Inertia"
+	TotpPeriod           = 30
+	TotpSecretSize       = 10
+	TotpDigits           = 6
+	TotpAlgorithm        = otp.AlgorithmSHA1
```
Comment (bfbachmann) [inline daemon/inertiad/crypto/totp.go]: Can we use a more secure hash function than SHA1 for this? As I understand it, SHA1 is considered broken for security purposes.

File: daemon/inertiad/crypto/totp.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+func generateSecretKey(accountName string) (*otp.Key, error) {
+	key, err := totp.Generate(totp.GenerateOpts{
+		Issuer:      TotpIssuerName,
+		AccountName: accountName,
+		Period:      TotpPeriod,
+		SecretSize:  TotpSecretSize,
+		Digits:      TotpDigits,
+		Algorithm:   TotpAlgorithm,
+	})
+	if err != nil {
```
Comment (bfbachmann) [inline daemon/inertiad/crypto/totp.go]: Minor: You don't need this `if`, you can just do
```suggestion
return totp.Generate(...)
```

File: daemon/inertiad/crypto/totp.go (hunk 4 above)
Comment (tristancalderbank) [reply daemon/inertiad/crypto/totp.go]: needs to be SHA-1 to be compatible with Google Authenticator and probably other TOTP apps

https://medium.com/@tilaklodha/google-authenticator-and-how-it-works-2933a4ece8c2

File: daemon/inertiad/crypto/totp.go (hunk 5 above)
Comment (tristancalderbank) [reply daemon/inertiad/crypto/totp.go]: fixed

**Pull request merged**